src/simd.rs

```diff
@@ -0,0 +1,198 @@
+    let bitmask = _mm_shuffle_epi8(bitmask_lookup, higher_nibbles);
+    let tmp = _mm_and_si128(bitset, bitmask);
+    let result = _mm_cmpeq_epi8(tmp, bitmask);
+    _mm_movemask_epi8(result)
+}
+
+/// Calls callback on bytes. Breaks when callback returns LoopInstruction::BreakAtWith(ix, val). And skips the
+/// number of bytes in callback return value otherwise.
+/// This method returns final index and a possible break value.
+#[cfg(all(target_arch = "x86_64", feature="simd"))]
+pub(crate) fn iterate_special_bytes<F, T>(bytes: &[u8], mut ix: usize, mut callback: F) -> (usize, Option<T>)
+    where F: FnMut(usize) -> LoopInstruction<Option<T>> 
+{
+    let upperbound = bytes.len().saturating_sub(15);
+    
+    while ix < upperbound {
+        let mut mask = unsafe { compute_mask(bytes, ix as isize) };
+        let block_start = ix;
+
+        while mask != 0 {
+            let mask_ix = mask.trailing_zeros() as usize;
+            ix += mask_ix;
+            match callback(ix) {
+                LoopInstruction::ContinueAndSkip(skip) => {
+                    ix += skip + 1;
+                    mask >>= skip + mask_ix + 1;
+                }
+                LoopInstruction::BreakAtWith(ix, val) => return (ix, val),
+            }
+        }
+
+        ix = std::cmp::max(ix, 16 + block_start);
+    }
+
+    let alignment = (bytes.as_ptr() as usize) % 16;
+    let remaining_bytes = bytes.len().saturating_sub(ix);
+
+    if remaining_bytes > 0 {
+        let mut mask = if alignment + remaining_bytes > 16 {
+            unsafe { compute_mask(bytes, ix as isize) }
+        } else {
+            // shift off the bytes we aren't interested in
+            unsafe { compute_mask(bytes, ix as isize - alignment as isize) >> alignment } 
+        };
+
+        // zero out the bytes we read past end of buffer
+        mask &= (1 << remaining_bytes) - 1;
+
+        while mask != 0 {
+            let mask_ix = mask.trailing_zeros() as usize;
+            ix += mask_ix;
+            match callback(ix) {
+                LoopInstruction::ContinueAndSkip(skip) => {
+                    ix += skip + 1;
+                    mask >>= skip + 1 + mask_ix;
+                }
+                LoopInstruction::BreakAtWith(ix, val) => return (ix, val),
+            }
+        }
+    }
+
+    (bytes.len(), None)
+}
+
+/// Scalar fallback.
+#[cfg(not(all(target_arch = "x86_64", feature="simd")))]
+pub(crate) fn iterate_special_bytes<F, T>(bytes: &[u8], mut ix: usize, mut callback: F) -> (usize, Option<T>)
+    where F: FnMut(usize) -> LoopInstruction<Option<T>> 
+{
+    while ix < bytes.len() {
+        match callback(ix) {
+            LoopInstruction::ContinueAndSkip(skip) => {
+                ix += skip + 1;
+            }
+            LoopInstruction::BreakAtWith(ix, val) => {
+                return (ix, val);
+            }
+        }
+    }
+
+    (ix, None)
+}
+
+#[cfg(all(target_arch = "x86_64", feature="simd", test))]
```

Comment:
I'm slightly uneasy the behavior is different in simd and non-simd cases, and this doesn't test the latter. But I think I see what you're doing.

Reply:
Agreed, this isn't great. I can try aligning the two without doing a double check on the non-simd case.